File: grade-school/grade_school.ml
(** solution using map *)
open Core.Std

module M = Int.Map

type school = string list M.t 

(** Create a new empty school *)
let create () = M.empty 

(** Add a student to a school *)
let add name g school = M.add_multi school ~key:g ~data:name

(** Get all the students from a grade *)
let grade g school = Option.value ~default:[] (M.find school g) 

(** Sort the list of students in a grade, if necessary *)
let sort school = M.map school ~f:(List.sort ~cmp:String.compare)

(** Return the list of grades and students as a map *)
let to_map school = school

File: fsharp/grade-school/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = Map.empty

let add (student: string) (grade: int) (school: School): School =
    match school.TryFind grade with
    | Some students -> school.Add (grade, List.sort (student :: students))
    | None -> school.Add (grade, [student])

let roster (school: School): (int * string list) list =
  Map.toList school |> List.sortBy   (fun (g, s) -> g) 

let grade (number: int) (school: School): string list =
  match school.TryFind number with
  | Some students -> students
  | None -> []

